### PR TITLE
Any-Order Magnus Expansion

### PR DESCRIPTION
**Pull request by ccoffrin**
WIP

### CODE REVIEW

**Comment by ccoffrin** [comment]
@zmorrell, I am done cleaning this up and it is ready for you to have a look before we merge it.  There are a lot of revisions in the algorithm section, the changelog has a summary of them.

File: src/dwave.jl [hunk 1]
```diff
@@ -4,29 +4,27 @@
...
 """
-function default_dwave_initial_state(n)
+function initial_state_default_dwave(n)
     return complex(ones(2^n)./(2^(n/2)))
 end
 
 
 """
-An AnnealingSchedule approximating those used in hardware by D-Wave Systems
+An AnnealingSchedule approx_IMATing those used in hardware by D-Wave Systems
```
Comment (zmorrell) [inline src/dwave.jl:13]: approx_IMATing should be approximating

File: src/simulate.jl [hunk 2]
```diff
@@ -100,64 +95,314 @@ ising_model - ising model represented as a dictionary.  The qubits
...
+
+"given to matricies, applies the commutator operation"
+function _matrix_commutator(a,b)
+    return a*b - b*a
+end
+
+"given two hamiltonians of orders 0-to-(n-1) and 0-to-(m-1), applies the _matrix_commutator operation"
+function _hamiltonian_commutator(h1::Vector, h2::Vector)
+    max_index = (length(h1)-1)+(length(h2)-1)+1
+    zeros = 0*(h1[1]*h2[2])
```
Comment (zmorrell) [inline src/simulate.jl:316]: if the length of h2 is < 2, this will break I think

**Review by zmorrell — COMMENTED** [review]
Pretty much everything looks good to me.  There is only one potential bug I found, and everything else is to make the code cleaner

File: src/simulate.jl [hunk 3]
```diff
@@ -16,75 +15,71 @@ const AS_QUADRATIC = AnnealingSchedule(s -> (1.0-s)^2, s -> s^2)
...
-        J_val = J[(i,j)]
-        sum_zs +=  zizj_vectorized(n, i, j, J_val)
-    end
-    return SparseArrays.spdiagm(sum_zs)
+function _sum_Z(n::Int, w::Vector)
+    return _tensor_sum_single_qubit(_ZMAT, n, w)
 end
 
-function sum_z_tup(n, tup, w)
+function _sum_Z(n::Int, t, w::Real)
```
[the remark this answers is not in the document]
Comment (ccoffrin) [reply src/simulate.jl]: changed to `_kron_z`

File: src/simulate.jl (hunk 2 above)
Comment (ccoffrin) [reply src/simulate.jl]: good point, it should have been `h2[1]`.

**Comment by ccoffrin** [comment]
great suggestions.  Thanks!

**Pull request merged**